build_runner_core/lib/src/asset/batch.dart

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:build/build.dart';
+import 'package:glob/glob.dart';
+import 'package:meta/meta.dart';
+
+import '../environment/io_environment.dart';
+import 'reader.dart';
+import 'writer.dart';
+
+/// Runs [inner] in a zone that indicates to compatible [RunnerAssetWriter]s
+/// that written assets should not be written to the underlying file system
+/// directly, but rather in a batch after [inner] has completed.
+///
+/// This solves a common issue when `build_runner` and other tools such as an
+/// analysis server are running concurrently: During a build, generated assets
+/// are written one-by-one. The analyzer will discover these changes while the
+/// build is running and report lints for a partial state of the workspace where
+/// not all generated files are present, leading to a subpar experience.
+///
+/// [runWithFileSystemBatch] is intended to wrap a single build run. Instead of
+/// writing assets directly, they are first cached in memory and then written at
+/// once after the build has completed. This causes fewer invalidations for
+/// external tools. The [writer] will be used to complete the writes collected
+/// while running [inner].
+///
+/// The underlying asset readers and writes both need to support file system
+/// batches (reader support is crucial to report changed assets that have not
+/// been flushed to the file system yet). The default [IOEnvironment] will use
+/// batch-aware readers and writers when not in low-resources mode.
+Future<T> runWithFileSystemBatch<T>(
+  Future<T> Function() inner,
+  RunnerAssetWriter writer,
+) async {
+  final batch = FileSystemWriteBatch._();
+  final result =
+      await runZoned(inner, zoneValues: {FileSystemWriteBatch._key: batch});
+  await batch._completeWrites(writer);
+  return result;
+}
+
+/// A batch of file system writes that should be committed at once instead of
+/// when [AssetWriter.writeAsBytes] or [AssetWriter.writeAsString] is called.
+///
+/// During a typical build run emitting generated files one-by-one, it's
+/// possible that other running tools such as an analysis server will have to
+/// re-analyze incomplete states multiple times.
+/// By storing pending outputs in memory first and then committing them at the
+/// end of the build, we have a better view over that needs to happen.
+///
+/// The default [IOEnvironment] uses readers and writes that are batch-aware
+/// outside of low-memory mode.
+@internal
+final class FileSystemWriteBatch {
+  final Map<AssetId, _PendingFileState> _pendingWrites = {};
+
+  FileSystemWriteBatch._();
+
+  Future<void> _completeWrites(RunnerAssetWriter writer) async {
+    await Future.wait(_pendingWrites.entries.map((e) async {
```

Comment:
It likely doesn't matter, but I recently did some benchmarking related to `.entries` and it is definitely slower than iterating the keys and then grabbing out the value manually.

I will leave it up to you, either is fine, I am sure the file I/O dwarfs this cost anyways.

Reply:
Counterintuitive, but I've changed it to iterate over keys with a separate lookup.